[GeoMechanicsApplication] Truss backbone law

WPK4FEM opened the pull request:
**📝 Description**
Addition of a backbone with elastic un- + reloading constitutive model for truss elements.

**🆕 Changelog**
- First setup of model
- Necessary fix of strain measure for linear truss element.

WPK4FEM committed:
First rudimentary version of backbone constitutive model for truss with test that shows unloading reloading.

WPK4FEM committed:
towards backbone on opposite side

WPK4FEM committed:
Test automation for tensile case

WPK4FEM committed:
Codesmells reduced

WPK4FEM committed:
Compressive side working and test provided.

WPK4FEM committed:
Simplified and fixed side choosing issue for backbone loading.

WPK4FEM committed:
Removed commented out code

WPK4FEM committed:
Removed unused function, added tests to suite.

WPK4FEM committed:
Looping tests and automation.

WPK4FEM committed:
unused variable removed

WPK4FEM committed:
Sonarcloud reported codesmells on python script for test automation removed.

WPK4FEM committed:
Sonarcloud. Rule of zero implemented.

WPK4FEM committed:
Sonarcloud. Unreachable return removed.

avdg81 committed:
Simplified the code a bit

- Moved some implementations from the header file to the implementation
  file.
- Reduced the scope of certain local variables.
- Added `[[nodiscard]]` attribute to some member functions.
- Removed the qualification from certain function calls.
- Removed pedantic `const` from member function parameters.
- Removed a few unused aliases.
- Pass to member function only data that is actually needed (i.e.
  provide Young's modulus rather than a bucket of data).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.h:
```diff
@@ -0,0 +1,93 @@
+// External includes
+
+// Project includes
+#include "includes/checks.h"
```
I believe this header file is intended to make checks available in unit tests. So it's unclear to my why we would need to include it here...?

AlejandroCornejo replied:
In CLs there is always a` Check` method to ensure that the input data is correct and sufficiently defined

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.h:
```diff
@@ -0,0 +1,93 @@
+    [[nodiscard]] double BackboneStress(double Strain) const;
+    [[nodiscard]] double BackboneStiffness(double Strain) const;
+    [[nodiscard]] double CalculateUnReLoadAmplitude(double YoungsModulus) const;
+    [[nodiscard]] bool IsWithinUnReLoading(double Strain, double YoungsModulus) const;
```
I think I forgot to reformat the file using clang-format the last time I've made changes to it. Could you please reformat it for me? Thank you.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+{
+ConstitutiveLaw::Pointer TrussBackboneConstitutiveLaw::Clone() const
+{
+    return Kratos::make_shared<TrussBackboneConstitutiveLaw>(*this);
```
Since this member function definition is contained in namespace `Kratos`, there is no need to repeat the namespace qualifier here:
```suggestion
    return make_shared<TrussBackboneConstitutiveLaw>(*this);
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+    rFeatures.mStrainSize = 1;
+
+    // Set the space dimension
+    rFeatures.mSpaceDimension = 3;
```
In my opinion, these comments repeat what the code already expresses clearly:
```suggestion
    rFeatures.mStrainSize = 1;
    rFeatures.mSpaceDimension = 3;
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+                                                            array_1d<double, 3>& rValue)
+{
+    KRATOS_ERROR << "Can't get the specified value for " << rThisVariable << std::endl;
+}
```
I don't understand the implementation of this member function. Are we supposed to never call it?

Also, when I was reading through the header file, I noticed this asymmetry in `GetValue` and `SetValue` overrides. There is a getter and setter for `Variable<double>` (which is good), but for `Variable<array_1d<double, 3>>` there only is the getter, but no setter. That confused me a bit. And now that I've seen this implementation, I'm even more confused. Could you please help me to understand the intention?

AlejandroCornejo replied:
IMO if you do not have anything to return, do not override it and even worse return an error. Imagine that you have 2 materials (this CL and another one in which printing the double var is relevant) in one simulation, if you try to print posts of a `double` type variable your code will crash because of this.

AlejandroCornejo replied:
Let me explain the reasoning at least within the SMApp:

- Each time you want to print anything at IP level, the` element.CalculatoOnIntegrationPoints()` is called.
- Then, this element method checks wether the `CL.Has()`  this variable (this method has to be implemented in the CL)
- If the `CL.Has()` is true then return the `CL.GetValue()`; if not return `CL.CalculateValue()`.
- If the CL is not related or does not know how to calculate something, it just does nothing, do not throw an error because of hat I said before.

AlejandroCornejo replied:
IDEM with `element.SetValuesOnIntegrationPoints()`

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+{
+    if (rThisVariable == TANGENT_MODULUS) {
+        const auto youngs_modulus = rParameterValues.GetMaterialProperties()[YOUNG_MODULUS];
+        const auto axial_strain(rParameterValues.GetStrainVector()[0]);
```
To be consistent with the previous line, I would rewrite this one as follows:
```suggestion
        const auto axial_strain = rParameterValues.GetStrainVector()[0];
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+void TrussBackboneConstitutiveLaw::CalculateMaterialResponsePK2(Parameters& rValues)
+{
+    // backbone loading, un- and reloading laws
+    const auto axial_strain(rValues.GetStrainVector()[0]);
```
In line with my previous suggestion:
```suggestion
    const auto axial_strain = rValues.GetStrainVector()[0];
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+            BackboneStress(mAccumulatedStrain + (std::abs(axial_strain - mUnReLoadCenter) -
+                                                 (CalculateUnReLoadAmplitude(youngs_modulus) / 2.)));
+        axial_stress_vector[0] = (axial_strain - mUnReLoadCenter) >= 0. ? axial_stress_vector[0]
+                                                                        : -axial_stress_vector[0];
```
Wouldn't it be simpler to write:
```suggestion
        if (axial_strain - mUnReLoadCenter < 0.) axial_stress_vector[0] *= -1.0;
```
?

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+void TrussBackboneConstitutiveLaw::FinalizeMaterialResponsePK2(Parameters& rValues)
+{
+    CalculateMaterialResponsePK2(rValues);
+    const auto axial_strain(rValues.GetStrainVector()[0]);
```
In line with my earlier comments:
```suggestion
    const auto axial_strain = rValues.GetStrainVector()[0];
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+
+void TrussBackboneConstitutiveLaw::FinalizeMaterialResponsePK2(Parameters& rValues)
+{
+    CalculateMaterialResponsePK2(rValues);
```
Just a question: wouldn't it be logical if this member function were called directly after `CalculateMaterialResponsePK2`? In other words, assuming that is the case (I need to check), why is it repeated here? Or am I overlooking something?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+{
+    KRATOS_CHECK(rMaterialProperties.Has(YOUNG_MODULUS))
+    // andere input voor backbone nalopen
+    return 0;
```
Just a friendly reminder that this needs to be checked prior to merging this branch to `master`.

Also, since this is an `override`, shouldn't we need to call the `Check` member function of its base class first?

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+double TrussBackboneConstitutiveLaw::BackboneStress(double Strain) const
+{
+    double young_modulus = 200.;
+    return 0.5 * young_modulus * Strain;
```
These hard-coded values probably need to be made more generic prior to merging this branch to `master`? Perhaps something like `return BackboneStiffness(youngs_modulus) * Strain;`?

avdg81 replied:
It now returns the interpolated value calculated by the stress-strain diagram (table).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+double TrussBackboneConstitutiveLaw::BackboneStiffness([[maybe_unused]] double Strain) const
+{
+    double young_modulus = 200.;
+    return 0.5 * young_modulus;
```
And the same thing here.

avdg81 replied:
It now returns the derivative calculated by the stress-strain diagram (table).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,166 @@
+
+double TrussBackboneConstitutiveLaw::CalculateUnReLoadAmplitude(double YoungsModulus) const
+{
+    return 2. * BackboneStress(mAccumulatedStrain) / YoungsModulus;
```
Just out of curiosity: is there also a name for this quantity without the factor 2? I'm asking because everywhere this member is being called, we first divide its value by 2. So multiplying and then dividing again by 2 seems redundant to me. Or am I overlooking something?

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -403,6 +404,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_VARIABLE( POISSON_UNLOADING_RELOADING )
 
+    KRATOS_REGISTER_VARIABLE( ACCUMULATED_STRAIN )
+    KRATOS_REGISTER_VARIABLE( TANGENT_MODULUS )
```
If I recall correctly, you said something like we wouldn't need this variable. If that is the case, I would suggest to remove this variable registration.

avdg81 replied:
Removed.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -132,6 +132,8 @@
 #include "custom_constitutive/small_strain_umat_2D_plane_strain_law.hpp"
 #include "custom_constitutive/small_strain_umat_3D_interface_law.hpp"
 #include "custom_constitutive/small_strain_umat_3D_law.hpp"
+#include "custom_constitutive/thermal_dispersion_law.h"
```
I'm not sure why this include is needed for this PR...?

avdg81 replied:
Removed.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application_variables.cpp:
```diff
@@ -79,6 +79,9 @@ KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS( LOCAL_RELATIVE_DISPLACEMENT_VECTOR )
 KRATOS_CREATE_VARIABLE( Matrix, LOCAL_PERMEABILITY_MATRIX )
 
+KRATOS_CREATE_VARIABLE( double, ACCUMULATED_STRAIN)
+KRATOS_CREATE_VARIABLE( double, TANGENT_MODULUS)
```
Perhaps this variable is no longer needed (see my previous comment)?

avdg81 replied:
Removed.

avdg81 commented on applications/GeoMechanicsApplication/tests/test_helper.py:
```diff
@@ -588,6 +590,7 @@ def element_integration_point_values_at_time(result_item_name, time, output_data
         else:
             element_results = [item["value"] for item in matching_item["values"]]
 
+
```
This additional blank line is probably not needed...?

avdg81 replied:
Removed.

avdg81 commented on applications/GeoMechanicsApplication/tests/test_truss_backbone_mat.py:
```diff
@@ -0,0 +1,125 @@
+        elongation to 4 m. The resulting force displacement curve should initially follow
+        the backbone, unload elastically, reload elastically, then follow the backbone further.
+        """
+        test_name    = r'truss_backbone_material\tension'
```
I would suggest to change this to a generic path (rather than the Windows-style path it is now):
```suggestion
        test_name    = 'truss_backbone_material/tension'
```

avdg81 replied:
Replaced by using `os.path.join`.

avdg81 commented on applications/GeoMechanicsApplication/tests/test_truss_backbone_mat.py:
```diff
@@ -0,0 +1,125 @@
+        # output
+        output_file_name = os.path.join(project_path, 'truss_backbone_mat.post.res')
+        reader = test_helper.GiDOutputFileReader()
+        output_data = []
```
Since there is always exactly one stage, I would suggest to not use a list, but simply an object that keeps the output data.  As a result, we no longer need variable `stage_nr`. This suggestion also applies to the other tests in this file.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/test_truss_backbone_mat.py:
```diff
@@ -0,0 +1,125 @@
+        for time, expected_force in zip(times, expected_forces):
+            # integration point check in element 1, integration point 1 ( uniform stress and strain so an arbitrary choice )
+            section_force = test_helper.GiDOutputFileReader.element_integration_point_values_at_time("FORCE", time, output_data[stage_nr], [1], [0])[0][0]
+            self.assertAlmostEqual(float(section_force[0]), expected_force, 2)
```
Shouldn't the values returned by `element_integration_point_values_at_time` already be `float`s? In other words, what is the reason we would need to explicitly cast to a `float` here?

avdg81 replied:
Removed the explicit type conversion.

avdg81 commented on applications/GeoMechanicsApplication/tests/truss_backbone_material/tension/ProjectParameters.json:
```diff
@@ -0,0 +1,148 @@
+            "variable_name":   "DISPLACEMENT",
+            "active":          [true,true, true],
+            "is_fixed":        [true,true,true],
+            "value":           [1.0E-02,0.0,0.0],
```
Does the first value here (1.0E-02) have any significance here, since we are using Table 1?

avdg81 replied:
The value has been changed to 0.0.

avdg81 commented on applications/GeoMechanicsApplication/tests/truss_backbone_material/compression/MaterialParameters.json:
Since this file is identical for all four test cases, I would suggest to move them into a `common` subdirectory.

avdg81 replied:
Done.

avdg81 reviewed (commented):
Thank you for making this excellent start. I will try to address my review feedback and to complete the missing parts.

avdg81 committed:
Avoid using Windows path separators explicitly in Python scripts

avdg81 committed:
Removed some redundant casts to `float` objects

The objects already are of type `float`.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/truss_backbone_law

avdg81 committed:
Generalized the backbone law's implementation and fixed some problems

Changes include:
- Defined, created and registered two vector variables for keeping the
  strains and stresses of piecewise linear constitutive laws.
- Added piecewise linear constitutive laws--consisting of a single line
  segment--to the material properties files of the tests.
- Reformatted some lines of the JSON files that contain the material
  properties.
- Made the calculation of the derivative outside the given range of a
  table consistent with how values are extrapolated.
- Replaced a usage of `KRATOS_CHECK` (which is intended for unit tests)
  by `KRATOS_ERROR_IF_NOT` (which is meant for detecting input/runtime
  errors).
- Added more input checks for the backbone constitutive law.
- The backbone stress is now properly calculated based on the given
  piecewise linear constitutive law. From the same input, we now also
  calculate the backbone stiffness. In other words, the adopted values
  are no longer hard-coded to satisfy the unit tests.
- The backbone constitutive law is now properly initialized. That is, we
  keep a table that describes the piecewise linear constitutive law.
- Implemented member function `Info` of the backbone constitutive law
  (for debugging purposes).
- The 2-noded 3D truss element from the structural mechanics application
  now initializes the material when the element is initialized. This is
  in line with other elements from that application.

avdg81 committed:
Processed some of my own review suggestions

- Reformatted a file using clang-format.
- Removed a redundant namespace qualifier.
- Removed a few comments that didn't add anything to what the code
  itself already expressed.
- Removed the `override` of `GetValue` that returned an
  `array_1d<double, 3>`, which raised an error when called.
- Rewrote a few variable initializations.
- Simplified changing a variable's value when a certain condition was
  met.

avdg81 committed:
Continued to process the review suggestions

- Also carry out any checks implemented by the base class.
- Avoid constantly adjusting the unload/reload amplitude by first
  multiplying by 2 and then dividing it by 2 again.
- Removed the creation and registration of variable `TANGENT_MODULUS`,
  since it wasn't being used.
- `output_data` in the integration tests for the truss backbone material
  is no longer a list, since there was always exactly one item in it. As
  a result, we could also remove variable `stage_nr`.
- Removed an `#include` that was accidentally added.
- Removed a blank line that was accidentally added.
- Changed four unused input values (which were overridden by values from
  a table) to 0.0.
- Removed the duplicated `MaterialParameters.json` files. Instead, we
  now have a single one that is used by all tests.
- Added a check for the backbone stiffness values.

avdg81 committed:
Completed the checks for the user-defined stress-strain diagram

- Check that the strain values are ascending.
- Reworked the checking code such that it becomes easier to read.

avdg81 committed:
Added some unit tests for the truss backbone law

avdg81 committed:
Added still more unit tests for the truss backbone law

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/truss_backbone_law

avdg81 committed:
Check the horizontal displacements in the integration tests

Also renamed a few local variables to be more precise.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/truss_backbone_law

avdg81 committed:
Fixed a compilation error

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/truss_backbone_law

avdg81 committed:
Removed a comment

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,223 @@
+    if (rThisVariable == ACCUMULATED_STRAIN) {
+        rValue = mAccumulatedStrain;
+    } else {
+        KRATOS_ERROR << "Can't get the specified value for " << rThisVariable << std::endl;
```
To be discussed with @rfaasse: shouldn't we call here `GetValue` of the base class rather than throwing an exception?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,223 @@
+    if (rThisVariable == ACCUMULATED_STRAIN) {
+        mAccumulatedStrain = rValue;
+    } else {
+        KRATOS_ERROR << "Can't set the specified value for " << rThisVariable << std::endl;
```
To be discussed with @rfaasse: shouldn't we call here `SetValue` of the base class rather than throwing an exception?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,223 @@
+            rValue = BackboneStiffness(axial_strain);
+        }
+    } else {
+        KRATOS_ERROR << "Can't calculate the specified value" << std::endl;
```
To be discussed with @rfaasse: shouldn't we call here `CalculateValue` of the base class rather than throwing an exception?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,223 @@
+                                                 CalculateUnReLoadAmplitude(youngs_modulus)));
+        if (axial_strain - mUnReLoadCenter < 0.0) axial_stress_vector[0] *= -1.0;
+    }
+    rValues.SetStressVector(axial_stress_vector);
```
@rfaasse: I doubt whether this call is needed. The corresponding `GetStressVector` returns a _reference_ to the stress vector. So assigning to it should already give the desired effect. `SetStressVector` just sets the pointer to the stress vector, which has not been changed here (we still point to the same object). Do you agree?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,223 @@
+{
+    ConstitutiveLaw::InitializeMaterial(rMaterialProperties, rElementGeometry, rShapeFunctionsValues);
+
+    const auto& r_strains_of_piecewise_linear_law = rMaterialProperties[STRAINS_OF_PIECEWISE_LINEAR_LAW];
+    const auto& r_stresses_of_piecewise_linear_law = rMaterialProperties[STRESSES_OF_PIECEWISE_LINEAR_LAW];
+    for (auto i = std::size_t{0}; i < r_strains_of_piecewise_linear_law.size(); ++i) {
+        mStressStrainTable.PushBack(r_strains_of_piecewise_linear_law[i],
+                                    r_stresses_of_piecewise_linear_law[i]);
+    }
```
@rfaasse : What do you think? Should we empty the table prior to filling it (to allow for this initialization to be executed more than once)?

avdg81 committed:
Added missing serialization code

avdg81 committed:
Improved getting, setting, and calculating values

Handle the specific cases first. If they don't apply, forward the
request to to the base class member functions.

avdg81 committed:
Fixed a code smell and removed a needless statement and comment

avdg81 committed:
Clear the table prior to populating it

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,228 @@
+
+double& TrussBackboneConstitutiveLaw::CalculateValue(ConstitutiveLaw::Parameters& rParameterValues,
+                                                     const Variable<double>&      rThisVariable,
+                                                     double&                      rValue)
```
It may be just `double Value` . No? rValue is an input and output at the same time.

avdg81 replied:
No, unfortunately we can't change that, since this member function is an `override`. So the function signature may not be changed. The difficulty with this inherited interface is that the result is returned through the parameter list (as you correctly noticed) as well as by using the return type.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,228 @@
+        auto pos_side = (axial_strain - mUnReLoadCenter) > 0.;
+        mAccumulatedStrain +=
+            std::abs(axial_strain - mUnReLoadCenter) - CalculateUnReLoadAmplitude(youngs_modulus);
+        mUnReLoadCenter = pos_side ? axial_strain - CalculateUnReLoadAmplitude(youngs_modulus)
```
I would merge `pos_side `into line 101. Or introduce a variable `deviation= axial_strain - mUnReLoadCenter` to be used in two places.

avdg81 replied:
Good observation. I have defined a new variable `const auto difference = axial_strain - mUnReLoadCenter;` and used that in the next two statements. The single usage of variable `pos_side` has been inlined.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,228 @@
+{
+    if (const auto exit_code = ConstitutiveLaw::Check(rMaterialProperties, rElementGeometry, rCurrentProcessInfo);
+        exit_code != 0)
+        return exit_code;
```
The current implementation of `ConstitutiveLaw::Check` (why not `BaseType::Check`?) returns 0. Is it worth to keep it?

avdg81 replied:
Yes, you are correct that at present it always returns 0. However, that is an implementation detail, which may change without notification in the future. Therefore, we decided to store and return the value returned by the base class `Check` function if it ever returns a non-zero value.

And yes, I agree that using `BaseType` here is actually better than repeating the name of the base class. I have found another similar occurrence in `InitializeMaterial`. I have changed that one, too, as well as three occurrences in the header file (in `using` statements). Thanks for pointing this out.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/truss_backbone_constitutive_law.cpp:
```diff
@@ -0,0 +1,228 @@
+    return mStressStrainTable.GetValue(Strain);
+}
+
+double TrussBackboneConstitutiveLaw::BackboneStiffness([[maybe_unused]] double Strain) const
```
@avdg81 , please could you clarify use of [[maybe_unused]]? As far as I understand it helps to avoid a warning when, for example, switching between Debug and Release compiler options. Thank you.

avdg81 replied:
Good catch! This was actually a left-over of @WPK4FEM first implementation, where he didn't use the `Strain` parameter. To suppress a warning, we then added this C++ attribute. But we had forgotten to remove it later, when we actually started to use the `Strain` parameter. It has been removed now.

From the [cppreference.com](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) website:
> For entities declared `[[maybe_unused]]`, if the entities or their structured bindings are unused, the warning on unused entities issued by the compiler is suppressed.

markelov208 reviewed (commented):
Hi @avdg81 and @WPK4FEM , nice work. I have just a few minor comments in addition to other comments. @avdg81 , I've found `std::adjacent_find`, is it possible to use it with ublas Vector? If yes, it can be used for ascending check. Thank you.

avdg81 committed:
Processed review suggestions

- Introduced a better intermediate variable.
- Use `BaseType` where appropriate rather than repeating the base
  class's name explicitly.
- Removed attribute `[[maybe_unused]]` that was not supposed to be there
  anymore.

avdg81 committed:
Replaced a raw loop by an algorithm

markelov208 reviewed (approved):
Hi Anne, thank you for quick changes and the algorithm! I shall copy it in my files for future use.
